### DISCUSSION TITLE
Update dependencies

### DISCUSSION THREAD

**Discussion by MartinHeroux** [doconce/doconce#62] [General]
There are a few places in the DocOnce repo where dependencies, recommendations or packages/programs are named/listed. Some are Python packages, others are not.

It would be nice to combine these lists and identify which dependencies are required, which are still being maintained, and which ones are required for which functionality.

The places I identified that should be considered are:
* setup.py
* doconce/requirements/
* doconce/doc/src/manual/ [various install scripts + manual itself]

# PYTHON 
Python v3.4 to v3.6 (manual still refers to v2.7, which is no longer supported)

__Update__ Noticed that .travis.yml file includes the following:

```
language: python
python:
  - "2.6"
  - "2.7"
  - "3.2"
  - "3.3"
  - "3.4"
  - "3.5"
  - "3.6"
```
Should reference to Python 2 be removed?

# PYTHON PACKAGES
It might prove useful to include as many of the required/recommended packages in setup.py, that way they are installed when DoConce is installed.

Three or four of the dependencies are hosted on the DocOnce github page. Would it be possible (or worthwhile) to publish these on PyPi so that they can be installed with pip (e.g. pip install publish)? Is it possible to have non-PyPi packages listed as dependencies that need to be installed when running setup.py?

Also, none of the recommended packages have a pinned version. Is this a problem in terms of dependencies possibly breaking if they adopt aspects of Python only available from v3.7 to 3.9? 

## DocOnce - required 
(Would it be possible to remove this package if there is no desire to stay 2.7 compatible?)
future

## DocOnce - recommended 
git+https://github.com/doconce/preprocess#egg=preprocess
mako

## to generate ipynb
ipython     **required?**
jupyter 		

## Publish - for references
git+https://github.com/doconce/publish#egg=publish
python-Levenshtein-wheels    # replacing python-Levenshtein
lxml			

## Pyments - syntax highlighting 
pygments
git+https://github.com/doconce/pygments-ipython-console#egg=pygments-ipython-console
git+https://github.com/doconce/pygments-doconce#egg=pygments-doconce

## Sphinx
sphinx

## Sphinx themes			
tinkerer
alabaster		
cloud_sptheme
sphinx_rtd_theme	
sphinxjp.themes.solarized
git+https://github.com/ryan-roemer/sphinx-bootstrap-theme#egg=sphinx-bootstrap-theme
git+https://github.com/shkumagai/sphinxjp.themes.impressjs#egg=sphinxjp.themes.impressjs
git+https://github.com/kriskda/sphinx-sagecell#egg=sphinx-sagecell

## Runestone sphinx books
sphinxcontrib-paverutils
paver
cogapp

## xhtml 
beautifulsoup4
html5lib
(what part of DocOnce uses these?)

## ptex2tex
git+https://github.com/doconce/ptex2tex#egg=ptex2tex  
**remove? depricate?**

"Originally, DocOnce relied on the ptex2tex program for very flexible choice of typesetting of verbatim code blocks. A simplified version, doconce ptex2tex, is bundled with DocOnce. However, even greater flexibility is now offered by the --latex_code_style= option to doconce format so unless you already are a ptex2tex user, it is recommended to forget about ptex2tex and just use the --latex_code_style= option."

**UPDATE**
I had a look at the ptex2tex repo and the last activity was in 2016 and Issues are turned off.

I tried to install the package as indicated in the install_doconce.sh (and the manual.pdf), but I was not able to because the package has not updated to Python 3 (e.g. setup.py includes old-style print statements).

I tried running bash.sh from test/SpringerT2/ and got the following error:

```
(venv) doconce@doconce:~/doconce/test/Springer_T2$ bash make.sh
running preprocess -DFORMAT=pdflatex -DDEVICE=paper -DPRIMER_BOOK -DCHAPTER="chapter" -DBOOK="book" -DAPPENDIX="appendix" -DALG="code" Springer_T2_book.do.txt > tmp_preprocess__Springer_T2_book.do.txt
running mako on tmp_preprocess__Springer_T2_book.do.txt to make tmp_mako__Springer_T2_book.do.txt
mako variables: {'FORMAT': 'pdflatex', 'DEVICE': 'paper', 'CHAPTER': 'chapter', 'BOOK': 'book', 'APPENDIX': 'appendix', 'PRIMER_BOOK': True, 'ALG': 'code'}
Translating doconce text in tmp_mako__Springer_T2_book.do.txt to pdflatex
open file with encoding utf-8
standalone exercises in Springer_T2_book_exercises.zip
found info about 2 exercises
Final all \\label\{(.+?)\}

exporting publish database papers.pub to papers.bib:
output in Springer_T2_book.p.tex
make.sh: line 6: ptex2tex: command not found
make.sh: unsuccessful command ptex2tex Springer_T2_book
abort!
```
By not having --latex_code_style= in the doconce command, the file that was output was a .p.tex file. When --latex_code_style= is used, the output is a .tex file and it is not necessary to run ptex2tex Springer_T2_book    

I added `--latex_code_style=lst` to the doconce command, which generated a .tex file. I could then remove the line in bash.sh that referred to ptex2tex: `ptex2tex Springer_T2_book` and was able to compile the book.

Also note that it is also possible to run `docconce ptex2tex Springer_T2_book`, which runs the built-in version of ptex2tex; but this seems to not be the recommended way to compile pdflatex.

**UPDATE**

Noted that in  `doconce/lib/doconce/doconce_config_default.py ` there is this item commented out: `#latex_code_style =`

Would it not be possible to added a default value and do away with ptex2tex?

## epydoc 
svn+https://epydoc.svn.sourceforge.net/svnroot/epydoc/trunk/epydoc#egg=epydoc 
**remove? depricate?**

"epydoc is an old-fashioned output format, will any doconce user use it?"

# LaTeX
(Worth highlighting its importance for any PDF output)
texinfo
texlive
texlive-extra-utils
texlive-latex-extra
texlive-latex-recommended
texlive-science
texlive-font-utils
texlive-fonts-extra
texlive-humanities
texlive-generic-extra
texlive-xetex
latexdiff
auctex

"Many LaTeX packages are potentially needed, depending on various constructions in the text and command-line options used when compling DocOnce to LaTeX. The standard packages always required are relsize, makeidx, setspace, color, amsmath, amsfonts, xcolor, bm, microtype, inputenc, and hyperref. Optional packages that might be included in the .tex output are minted, listings, fancyvrb, xunicode, inputenc, helvet, mathpazo, wrapfig, calc, ifthen, xkeyval, tikz, graphicx, setspace, shadow, disable, todonotes, lineno, xr, framed, mdframe, movie15, a4paper, and a6paper.

Relevant Debian packages that gives you all of these LaTeX packages are
texlive
texlive-extra-utils
texlive-latex-extra
texlive-font-utils

Alternatively, one may pull in texlive-full to get all available style files."

# VERSION CONTROL
(Not explicitly required, but is needed to install Python packages not on PyPi and thus can't be installed using pip)

mercurial	  **remove?**
git
subversion    **remove?**

# IMAGE MANIPULATION
(Not explicitly required, but used by DocOnce) 
imagemagick

"DocOnce calls up programs from  the  ImageMagick  suite  for  converting  image  files  to  a  proper  format  if needed.  The ImageMagick suite can be installed on all major platforms."

# MARKDOWN AND PANDOC OUTPUT
(Not explicitly required, but used by DocOnce) 
pandoc

"The DocOnce format pandoc outputs the document in various Markdown versions: the Pandoc extended Markdown format (which via the pandoc program can be translated to a range of other formats), strict Markdown, and GitHub-flavored Markdown."

# DEBIAN PACKAGES

build-essential

## Related to Publish
libxml2-dev # Development files for the GNOME XML library
libxslt1-dev # XSLT 1.0 processing library - development kit
zlib1g-dev  # compression library - development

## Image manipulation
inkscape
netpbm
mjpegtools
pdftk  # dropped by Debian; snap install 
giftrans
gv
evince
smpeg-plaympeg
mplayer
totem
ffmpeg

## Misc
ispell    # for spellchecking
aspell   # for spellchecking
libreoffice
unoconv
libreoffice-dmaths

## diff tools
curl
a2ps
wdiff
meld
diffpdf
kdiff3
diffuse

**Comment by MartinHeroux** [comment]
I looked into what packages are needed for the various aspects of DocOnce. I have tried to reduce the number of dependencies.

At present , I would consider deprecating the use of ptex2tex (not installable on Python 3) and epydoc (not installable on Python 3, most likely never used by DocOnce users, requires svn)

# System requirements

System Package  |  Required/Optional  | Version  |  Functionality    
-----------------------  | ------------------------- | ----------  |   ----------------  
python                   |     required              |   <= 3.6 |   Doconce-base    
git              | required              |      | Doconce-base    
make         |     optional           |  |       Sphinx 
imagemagick | optional | | image manipulation      
texlive          |   optional            |  | pdflatex, Sphinx  
texlive-extra-utils |      optional  |       |      pdflatex, Sphinx  
texlive-latex-extra |      optional  |       |      pdflatex, Sphinx  
texlive-science   |      optional     |    |      pdflatex, Sphinx  
 texlive-font-utils   |    optional      |     |    pdflatex, Sphinx  
latexmk          |   optional         |   |        Sphinx       
ispell/aspell | optional | | spellchecking

# Python requirements

Python package   |    Available on PyPi | Required/Optional | version   |   Functionality       
------------------------- | ----------------- | ----------------- | ------- | ----------------------  
doconce           |    yes (old)     |    required  |   1.5.2   |     Doconce-base       
preprocess | yes (old) | required |  | Doconce-base preprocessing
sphinx            |      yes         |     optional     |           |       sphinx          
matplotlib         |       yes       |       optional   |           |         sphinx          
ipython            |      yes        |      optional    |          |         ipynb           
jupyter            |      yes          |    optional    |          |         ipynb           
sphinx_rtd_theme      |       yes      |        optional |        |           Sphinx-theme       
cloud_sptheme          |     yes        |      optional  |         |         Sphinx-theme       
sphinxjp.themes.solarized    |     yes    |          optional  |  |                Sphinx-theme       
sphinx-bootstrap-theme       |    no      |        optional    |  |              Sphinx-theme       
sphinxjp.themes.impressjs    |      no     |         optional  |  |                Sphinx-theme       
sphinx-sagecell             |  no         |     optional      |   |         sphinx-sagecell      
Sphinxcontrib-paverutils     |    yes     |         optional  |  |           Sphinx-Runestone books  
cogapp                 | yes           |   optional          |   |  Sphinx-Runestone books  
pygments-doconce          |    no      |        optional     |  |             code markup        
Pygments-ipython-console    |      no   |           optional  |   |               code markup        
python-Levenshtein-wheels    |     yes         |     optional |  |            Publish – references   
lxml                |   yes        |      optional            |  |  Publish – references   
publish         |          no        |      optional          | |     Publish – references

**Comment by aless80** [comment]


**Reply by MartinHeroux** [reply]
In terms of the Python packages that are not on PyPi, three are Sphinx themes and three are hosted on github on the DocOnce account.

Would it be possible to publish the three DocOnce-account packages to PyPi, and also push a new version of preprocess (the version on PyPi is old)?

In terms of the Sphinx themes, DocOnce [already ships with various themes](https://github.com/doconce/doconce/tree/master/bundled) for various output formats. I realise we don't want to continue adding content to DocOnce, but we could add the three Sphinx themes currently not on PyPi with the other themes already shipped with PyPi.

In this way, users could create a dedicated Python 3.6 venv and simply `pip install doconce`. This would also remove a hard system dependency: git.

As an example, installation instruction on Ubuntu 20.04 might look something like this:
```
# Install Python 3.6 with virtualenv
sudo add-apt-repository ppa:deadsnakes/ppa
sudo apt-get update
sudo apt-get install -y python3.6
sudo apt-get install -y python3-virtualenv

# Create Python venv
mkdir ~/doconce_projects
cd ~/doconce_projects
virtualenv -p /usr/bin/python3.6 venv
source venv/bin/activate

# Install doconce and its Python dependencies
pip install doconce
``` 

We could then mention the other optional system dependencies: imagemagick, LaTeX-related packages, spell-check library.
